File: 2021/day2/day2.clj
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2021/day2/input.txt")
   (str/split-lines)
   (map (fn [x]
          (-> x
              (str/split #" ")
              (update-in [1] #(Integer/parseInt %)))))))

(defn part1 [xs]
  (let [run-command (fn [result [dir num]]
                      (case dir
                        "forward" (update result :x #(+ % num))
                        "down" (update result :y #(+ % num))
                        "up" (update result :y #(- % num))))]
    (->> xs
         (reduce run-command {:x 0 :y 0})
         ((juxt :x :y))
         (apply list)
         (apply *))))

(defn part2 [xs]
  (let [run-command (fn [result [dir num]]
                      (case dir
                        "forward" (-> result
                                      (update :x #(+ % num)) 
                                      (update :y #(+ % (* num (:aim result)))))
                        "down" (update result :aim #(+ % num))
                        "up" (update result :aim #(- % num))))]
    (->> xs
         (reduce run-command {:x 0 :y 0 :aim 0})
         ((juxt :x :y))
         (apply list)
         (apply *))))

(println "day2 part1:" (part1 input))
(println "day2 part2:" (part2 input))

File: 2021/day2/day2.js
const fs = require('fs');
const path = require('path');

const COMMAND_REGEX = /^(\w+) (\d{1,})/;

const part1 = input => {
  const commands = input
    .trim()
    .split('\n')
    .map(s => s.match(COMMAND_REGEX));

  let x = 0;
  let y = 0;
  for (let [, dir, numStr] of commands) {
    switch (dir) {
      case 'forward':
        x += parseInt(numStr, 10);
        break;
      case 'down':
        y += parseInt(numStr, 10);
        break;
      case 'up':
        y -= parseInt(numStr, 10);
        break;
      default:
        break;
    }
  }
  return x * y;
};

const part2 = input => {
  const commands = input
    .trim()
    .split('\n')
    .map(s => s.match(COMMAND_REGEX));

  let x = 0;
  let y = 0;
  let aim = 0;
  for (let [, dir, numStr] of commands) {
    switch (dir) {
      case 'forward':
        let num = parseInt(numStr, 10);
        x += num;
        y += num * aim;
        break;
      case 'down':
        aim += parseInt(numStr, 10);
        break;
      case 'up':
        aim -= parseInt(numStr, 10);
        break;
      default:
        break;
    }
  }
  return x * y;
};

let input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');
// input = `forward 5
// down 5
// forward 8
// up 3
// down 8
// forward 2`
console.log('day2 part1:', part1(input));
console.log('day2 part2:', part2(input));
